### ISSUE TITLE
C++ performance and optimization

### ISSUE THREAD

**Issue by wustus** [ml-explore/mlx#1767]
I've initially implemented a neural network in C++, following Nielsens Book 'Neural Networks and Deep Learning' and using Eigen as a BLAS library.
After reimplementing the same network for different datasets and reaching one where the input was larger than before (3.072x1), one epoch trained for ~30 seconds, I wanted to look into GPU training. After some research I found MLX and switched out the Eigen Matrices for MLX Vectors.
Doing this for a smaller project (kmnist) with 28x28 grayscale images (same as MNIST), I get extremely slow code compared to Eigen.

My current code can be found here: https://github.com/wustus/kuzushiji

I'm trying to chain my operations and evaluate every few batches.
When trying to refactor for example the sigmoid_prime method and compile it for a speedup I got a Segmentation fault.
I know I'm probably better off using the MLX methods for SGD and I guess `Module` to create the actual network but I'd like to use it as a BLAS library to learn as much as I can about the underlying techniques.

**Comment by angeloskath** [comment]
Going briefly over the code I would say that the two things that stick out are
1. Doing the forward matmuls on the GPU but everything else on the CPU adding many synchronizations while not a lot of work needs to be done in the meantime
2. Similar to the above you have two more streams that are used in some computations occasionally. You don't need those, they are simply adding synchronization points especially problematic given that the operations that you are performing are tiny so they would benefit from pipelining.
3. Evaluating each array individually instead of all of them at once.

So my advice would be to remove all streams and simply set the default stream to the device you want to use and change the loop -> eval to a single call to `eval(...)`.

There are many more things that should be different like batching your computations instead of looping manually over the batch size and using function transformations like `grad` or `vmap` to do that without adding complexity. However, I get the feeling that this is not quite what you want to do and you simply want to use the `mx::array` kind of like a numpy or Eigen replacement.

**Comment by wustus** [comment]
Thanks for taking the time to look through my code.
I've tried to follow your suggestions: 
 - setting the GPU as my default stream
 - dropping the arbitrary streams in my `mlx` calls 
 - running `eval({...})` on the entire list of weights and biases instead of calling `.eval()` on each respectively

I fear though without using `mlx` the way it's 'supposed to be used' I won't achieve the speedup I was hoping for (but ~40% is still significant).
I added some measurements before and after the optimizations to the repository along with the Eigen implementation I used as a reference. 

Again, thank you for taking the time. I appreciate any help towards becoming a better dev tremendously.

From my side, this issue can be closed but I'll leave the door open for discussion.

**Comment by angeloskath** [comment]
Using batching would make a huge difference while you can still compute the gradients yourself. Basically there is no reason to loop over every sample, compute the gradient and `mx::add` it to the previous gradients. You can simply concatenate the inputs, compute all outputs at once and all gradients at once using a single matmul to do the aggregation which will be significantly faster. Much more than 40% :-) .

I am closing the issue but feel free to open a different one if you encounter any bug or unexpected behavior.
